src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.CompilationManager.cs

```diff
@@ -0,0 +1,218 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Diagnostics.EngineV2
+{
+    internal partial class DiagnosticIncrementalAnalyzer
+    {
+        /// <summary>
+        /// This cache CompilationWithAnalyzer for active/open files. 
+        /// This will aggressively let go cached compilationWithAnalyzers to not hold them into memory too long.
+        /// </summary>
+        private class CompilationManager
+        {
+            private readonly DiagnosticIncrementalAnalyzer _owner;
+            private ConditionalWeakTable<Project, CompilationWithAnalyzers> _map;
```

Comment:
We would need perf testing for this - I had taken this approach in the initial implementation of v2 analyzer driver and this caused VM regressions in RPS runs. I see that you have some aggressive cache cleanup logic here, so hopefully it wont be an issue.